### PR TITLE
Delete Ort Allocator in LearningModelBinding

### PR DESCRIPTION
**Pull request by ryanlai2**
It is an issue to delete OrtAllocators in ImageFeatureValue because LearningModelBinding is the object that holds on the Ort::Value

### CODE REVIEW

**Comment by walrusmcd** [comment]
>   (WinML::BindingContext& context, OrtValue** ort_value);

make the allocator an out param

---
Refers to: winml/lib/Api/ImageFeatureValue.h:36 in ae47af6. [](commit_id = ae47af67d95a49df668d5730dd1b34b1d1cf9a5c, deletion_comment = False)

File: winml/lib/Api/LearningModelBinding.h [hunk 1]
```diff
@@ -77,8 +77,10 @@ struct LearningModelBinding : LearningModelBindingT<LearningModelBinding, ILearn
   com_ptr<winmla::IWinMLAdapter> adapter_;
   std::vector<std::string> input_names_;
   std::vector<Ort::Value> inputs_;
+  std::vector<OrtAllocator*> input_allocators_;
```
Comment (walrusmcd) [inline winml/lib/Api/LearningModelBinding.h]: if we had smart pointers for these.... Ort::Allocator - as this would get us closer to where we want to be with the C_ABI as having allocators from ORT be real things and freeable.    But .. since we don't have the the C ABI yet you would have to make a smart pointer Ort::Allocator in one of our headers .. like the adapter, you can't put this in the cxx_api.h yet..

Comment (ryanlai2) [reply winml/lib/Api/LearningModelBinding.h]: Cool! I made an Ort::Allocator in the WinMLAdapter.h. It is based off of Base struct in onnxruntime_cxx_api.h

File: winml/adapter/WinMLAdapter.h [hunk 2]
```diff
@@ -146,4 +146,60 @@ class InferenceSession : public Microsoft::WRL::RuntimeClass <
...
+    v.m_ort_allocator = nullptr;
+  }
+  void operator=(Allocator&& v) noexcept {
+    if (m_ort_allocator != nullptr && m_adapter != nullptr) {
+      m_adapter->FreeProviderAllocator(m_ort_allocator);
+    }
+    m_adapter = v.m_adapter;
+    m_ort_allocator = v.m_ort_allocator;
+    v.m_adapter = nullptr;
+    v.m_ort_allocator = nullptr;
```
Comment (tiagoshibata) [inline winml/adapter/WinMLAdapter.h]: I think these 4 lines can be replaced with `this = std::move(v)` (or just `this = v`, since it's an rvalue already?) to avoid duplicating the move code.

File: winml/lib/Api/LearningModelBinding.cpp [hunk 3]
```diff
@@ -21,6 +21,11 @@ LearningModelBinding::LearningModelBinding(
 }
 WINML_CATCH_ALL
 
+LearningModelBinding::~LearningModelBinding()
+{
+  Clear();
+}
+
```
Comment (tiagoshibata) [inline winml/lib/Api/LearningModelBinding.cpp]: Why is this needed?

File: winml/lib/Api/ImageFeatureValue.cpp [hunk 4]
```diff
@@ -519,13 +513,14 @@ HRESULT ImageFeatureValue::GetOrtValue(WinML::BindingContext& context, OrtValue*
   OrtAllocator* dml_allocator;
   WINML_THROW_IF_FAILED(m_adapter->GetProviderAllocator(provider, &dml_allocator));
 
+  std::unique_ptr<OrtAllocator>ort_allocator_temp(dml_allocator);
```
Comment (tiagoshibata) [inline winml/lib/Api/ImageFeatureValue.cpp]: Why is ort_allocator_temp needed? It's constructing a unique_ptr to OrtAllocator from dml_allocator, which is a pointer to OrtAllocator, and then calling release(). You could use dml_allocator directly, right?

File: winml/lib/Api/LearningModelBinding.cpp [hunk 5]
```diff
@@ -121,9 +127,10 @@ std::tuple<std::string, OrtValue*, BindingType> LearningModelBinding::CreateBind
   if (!isPlaceHolder || shouldAlwaysTensorize) {
     // If not a placeholder, attempt to get the underlying resource
     WINML_THROW_IF_FAILED_MSG(
-        spLotusValueProvider->GetOrtValue(context, value.put()),
+        spLotusValueProvider->GetOrtValue(context, value.put(), ort_allocator.put()),
         "The model variable %s failed tensorization.",
         name.c_str());
+    auto spImageFeatureValue = spLotusValueProvider.try_as<ImageFeatureValue>();
```
Comment (tiagoshibata) [inline winml/lib/Api/LearningModelBinding.cpp]: Unused variable

File: winml/lib/Api/LearningModelBinding.cpp [hunk 6]
```diff
@@ -611,7 +626,8 @@ void LearningModelBinding::BindUnboundOutputs() {
   // Add all unbound outputs to binding collection
   for (const auto& unbound_output : unbound_output_names) {
     Ort::Value out(nullptr);
-    WINML_THROW_IF_FAILED(BindOutput(unbound_output, out));
+    Ort::Allocator out_allocator(adapter_.get(), nullptr);
+    WINML_THROW_IF_FAILED(BindOutput(unbound_output, out, out_allocator));
```
Comment (tiagoshibata) [inline winml/lib/Api/LearningModelBinding.cpp]: A few comments:

* BindOutput and BindInput are calling release() on the allocator. Does this loop work for multiple outputs? On the second iteration you're passing a null allocator, right?
  * If you didn't want BindOutput to modify the allocator, you can make the argument a `const Ort::Allocator&` so the compiler enforces that you're not changing it
* Why cache the adapter in Ort::Allocator and not use it? Every time you wans an allocator, you're releasing the old one and calling adapter_.get(). You could use move semantics instead since you created nice definitions for `Allocator(Allocator&& v)` and `void operator=(Allocator&& v)` above. The code will probably be prettier and you won't have to call adapter_.get() every time you want an allocator.

File: winml/lib/Api/ImageFeatureValue.cpp (hunk 4 above)
Comment (tiagoshibata) [reply winml/lib/Api/ImageFeatureValue.cpp]: And it's dangerous, if you this function returns or throws before you call release() it would try to `delete` the pointer, which is undefined behavior (it must call `m_adapter->FreeProviderAllocator` instead of `delete`)

Comment (ryanlai2) [reply winml/lib/Api/ImageFeatureValue.cpp]: good point! I think I should use the Ort::Allocator I created instead

---
In reply to: [358565189](https://github.com/microsoft/onnxruntime/pull/2653#discussion_r358565189) [](ancestors = 358565189)

File: winml/adapter/WinMLAdapter.h (hunk 2 above)
Comment (ryanlai2) [reply winml/adapter/WinMLAdapter.h]: I don't think it's good idea to use std code in header files shared between two DLLs. Let me try and see if I can avoid duplicating the code.

---
In reply to: [358555221](https://github.com/microsoft/onnxruntime/pull/2653#discussion_r358555221) [](ancestors = 358555221)

File: winml/lib/Api/LearningModelBinding.cpp (hunk 3 above)
Comment (ryanlai2) [reply winml/lib/Api/LearningModelBinding.cpp]: not needed! I was using this earlier when I didn't have Ort::Allocator

---
In reply to: [358556189](https://github.com/microsoft/onnxruntime/pull/2653#discussion_r358556189) [](ancestors = 358556189)

File: winml/adapter/WinMLAdapter.h (hunk 2 above)
Comment (tiagoshibata) [reply winml/adapter/WinMLAdapter.h]: Ok, I think `this = v` works and invokes the move assignment you defined. `std::move` shouldn't pose problems (it's just a cast and shouldn't matter for ABI compatibility) but it's really better to avoid STL altogether

File: winml/lib/Api/LearningModelBinding.cpp (hunk 3 above)
Comment (ryanlai2) [reply winml/lib/Api/LearningModelBinding.cpp]: Actually this is needed to ensure that Ort::Allocator destructs after Ort::Value

---
In reply to: [358556189](https://github.com/microsoft/onnxruntime/pull/2653#discussion_r358556189) [](ancestors = 358556189)
